sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -165,43 +216,37 @@ class GpuSingleDirectoryDataWriter(
 
   override def write(batch: ColumnarBatch): Unit = {
     val maxRecordsPerFile = description.maxRecordsPerFile
-    if (!needSplitBatch(maxRecordsPerFile, recordsInFile, batch.numRows())) {
+    val dataTypes = GpuColumnVector.extractTypes(batch)
+    if (!shouldSplitToFitMaxRecordsPerFile(maxRecordsPerFile, recordsInFile, batch.numRows())) {
       closeOnExcept(batch) { _ =>
         statsTrackers.foreach(_.newBatch(currentWriter.path(), batch))
         recordsInFile += batch.numRows()
       }
       currentWriter.writeAndClose(batch, statsTrackers)
     } else {
-      withResource(batch) { batch =>
-        withResource(GpuColumnVector.from(batch)) {table => 
-          val splitIndexes = getSplitIndexes(
-            maxRecordsPerFile,
-            recordsInFile,
-            table.getRowCount()
-          )
-  
-          val dataTypes = GpuColumnVector.extractTypes(batch)
-  
-          var needNewWriter = recordsInFile >= maxRecordsPerFile
-          withResource(table.contiguousSplit(splitIndexes: _*)) {tabs =>
-            tabs.foreach(b => {
-              if (needNewWriter) {
-                fileCounter += 1
-                assert(fileCounter <= MAX_FILE_COUNTER,
-                  s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
-                newOutputWriter()
-              }
-              withResource(b.getTable()) {tab =>
-                val bc = GpuColumnVector.from(tab, dataTypes)
-                closeOnExcept(bc) { _ =>
-                  statsTrackers.foreach(_.newBatch(currentWriter.path(), bc))
-                  recordsInFile += b.getRowCount()
-                }
-                currentWriter.writeAndClose(bc, statsTrackers)
-                needNewWriter = true
-              }
-            })
+      val partContigTables = withResource(batch) { batch =>
+        withResource(GpuColumnVector.from(batch)) { table =>
+          splitToFitMaxRecords(table, maxRecordsPerFile, recordsInFile)
+        }
+      }
+      var needNewWriter = recordsInFile >= maxRecordsPerFile
+      closeOnExcept(partContigTables) { _ =>
+        partContigTables.zipWithIndex.foreach { case (partContigTable, partIx) =>
+          if (needNewWriter) {
+            fileCounter += 1
+            assert(fileCounter <= MAX_FILE_COUNTER,
+              s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
+            newOutputWriter()
+          }
+          // null out the entry so that we don't double close
+          partContigTables(partIx) = null
+          withResource(partContigTable) { _ =>
+            val partBatch = GpuColumnVector.from(partContigTable.getTable, dataTypes)
+            statsTrackers.foreach(_.newBatch(currentWriter.path(), partBatch))
+            recordsInFile += partBatch.numRows()
+            currentWriter.writeAndClose(partBatch, statsTrackers)
```

Comment:
`writeAndClose` goes out of its way to close the batch early during processing, so any time we see a pattern like this:
```
withResource(x) {
  ...
  ...writeAndClose(..)
}
```
then something is probably wrong, because we're artificially holding onto something until the write of a batch completes, and a large part of the write is not on the GPU.  We need to close everything related to a batch except the batch itself, and then hand the batch to the writer so it can close it during the write processing at the appropriate time.